evolver/controller/standard/chemostat.py

```diff
@@ -0,0 +1,53 @@
+import time
+from pydantic import Field
+from collections import defaultdict, deque
+from evolver.controller.interface import Controller
+from evolver.hardware.interface import VialConfigBaseModel
+
+
+class Chemostat(Controller):
+    class Config(VialConfigBaseModel):
+        od_sensor: str = Field(description="name of OD sensor to use")
+        pump: str = Field(description="name of pump device to use")
+        stirrer: str = Field(description="name of stirrer device to use")
+        window: int = Field(7, description="number of OD measurements to collect prior to start")
+        start_od: float = Field(0, description="OD at which to start chemostat dilutions")
+        start_time: int = Field(0, description="Time (in hours) after which to start dilutions")
+        flow_rate: float = Field(0, description="Flow rate for dilutions")
+        stir_rate: float = Field(8, description="Stir rate")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # buffer could come from history as well
+        self.od_buffer = defaultdict(lambda: deque(maxlen=self.config.window))
+        # start_time is something we might actually want to be a property of
+        # evolver, in case of interrupt it has a chance of continuing - but can
+        # come from history similarly
+        self.start_time = time.time()
+
+    def control(self, *args, **kwargs):
+        od_val = self.evolver.hardware.get(self.config.od_sensor).get()
+        pump = self.evolver.hardware.get(self.config.pump)
+        stir = self.evolver.hardware.get(self.config.stirrer)
```

Comment:
These are just aliases due to the long code and reuse of them. Could make them property, but they are also really only needed in the control and not meant to be provided, so don't see a huge motivation right now.

Reply:
This really took me a while to grapple with what your point might be - do you mean that these are _just_ aliases because you haven't tethered them anywhere else in the class and that's so by design?

These really should not be _just_ aliases. The control class's purpose is to encapsulate and contain specific control of specific hardware and that hardware should very much be significantly more bound to the controller implementation and instance than just having a single method, ``control``, have weak ref'd access to it.

The class would be significantly more extensible if the hardware was tethered as an instance attribute, e.g., ``self.pump``. Or if we want to keep them dynamic, or to also provide a useful hook, make them a property to abstract the retrieval method.

>and not meant to be provided

By that do you mean that they're not meant to be provided externally to the class, i.e., as part of its interface? It's totally a legit pattern to use properties as access refs (even if only used internally) when you want to abstract the access methodology.
 
Also, how would this work with the pattern of passing in hardware or having the controller instantiated from a descriptor?

Reply:
I made these into properties